I see this error, which also tells you how to fix it:
```
java.lang.AssertionError: key 'GroupByAllRequiresExplicitSelectList' not found for resource 'groupByAllRequiresExplicitSelectList' in bundle 'java.util.PropertyResourceBundle@4b5c5087'; add the following line to org.apache.calcite.runtime.CalciteResource.properties:
--
  | GroupByAllRequiresExplicitSelectList=GROUP BY ALL requires an explicit SELECT list; '*' is not supported
```

--- a/core/src/main/codegen/templates/Parser.jj
+++ b/core/src/main/codegen/templates/Parser.jj
@@ -2792,25 +2792,42 @@ SqlNode Where() :
 SqlNodeList GroupBy() :
 {
     final List<SqlNode> list;
-    final boolean distinct;
     final Span s;
+    SqlParserPos pos;
 }
 {
     <GROUP> { s = span(); }
     <BY>
     (
-        <DISTINCT> { distinct = true; }
-    |   <ALL> { distinct = false; }
-    |   { distinct = false; }
+        <DISTINCT>
+        list = GroupingElementList() {
+            pos = s.end(this);
+            return new SqlNodeList(
+                ImmutableList.of(
+                    SqlInternalOperators.GROUP_BY_DISTINCT.createCall(pos, list)), pos
+            );
+        }
+    |
+        <ALL>
+        (
+            LOOKAHEAD(2)
+            list = GroupingElementList() {
+                return new SqlNodeList(list, s.end(this));
+            }
+        |
+            {
+                pos = s.end(this);
+                return new SqlNodeList(
+                    ImmutableList.of(
+                        SqlInternalOperators.GROUP_BY_ALL.createCall(pos)), pos
+                );
+            }
+        )
+    |
+        list = GroupingElementList() {
+            return new SqlNodeList(list, s.end(this));
+        }
     )
-    list = GroupingElementList() {
-        final SqlParserPos pos = s.end(this);
-        final List<SqlNode> list2 = distinct
-            ? ImmutableList.of(
-                SqlInternalOperators.GROUP_BY_DISTINCT.createCall(pos, list))
-            : list;
-        return new SqlNodeList(list2, pos);
-    }
 }
 
 List<SqlNode> GroupingElementList() :

--- a/core/src/main/java/org/apache/calcite/runtime/CalciteResource.java
+++ b/core/src/main/java/org/apache/calcite/runtime/CalciteResource.java
@@ -383,6 +383,9 @@ public interface CalciteResource {
   @BaseMessage("Windowed aggregate expression is illegal in {0} clause")
   ExInst<SqlValidatorException> windowedAggregateIllegalInClause(String a0);
 
+  @BaseMessage("GROUP BY ALL requires an explicit SELECT list; ''*'' is not supported")
+  ExInst<SqlValidatorException> groupByAllRequiresExplicitSelectList();
+
   @BaseMessage("Aggregate expressions cannot be nested")
   ExInst<SqlValidatorException> nestedAggIllegal();
 

--- a/core/src/main/java/org/apache/calcite/sql/SqlKind.java
+++ b/core/src/main/java/org/apache/calcite/sql/SqlKind.java
@@ -183,6 +183,9 @@ public enum SqlKind {
   /** The DISTINCT keyword of the GROUP BY clause. */
   GROUP_BY_DISTINCT,
 
+  /** The ALL keyword of the GROUP BY clause. */
+  GROUP_BY_ALL,
+
   /**
    * ORDER BY clause.
    *

--- a/core/src/main/java/org/apache/calcite/sql/SqlSelectOperator.java
+++ b/core/src/main/java/org/apache/calcite/sql/SqlSelectOperator.java
@@ -214,6 +214,10 @@ public class SqlSelectOperator extends SqlOperator {
         writer.sep("GROUP BY DISTINCT");
         List<SqlNode> operandList = ((SqlCall) groupBy.get(0)).getOperandList();
         groupBy = new SqlNodeList(operandList, groupBy.getParserPosition());
+      } else if (groupBy.size() == 1 && groupBy.get(0) != null
+          && groupBy.get(0).getKind() == SqlKind.GROUP_BY_ALL) {
+        writer.sep("GROUP BY ALL");
+        groupBy = new SqlNodeList(groupBy.getParserPosition());
       } else {
         writer.sep("GROUP BY");
       }

--- a/core/src/main/java/org/apache/calcite/sql/fun/SqlInternalOperators.java
+++ b/core/src/main/java/org/apache/calcite/sql/fun/SqlInternalOperators.java
@@ -203,6 +203,11 @@ public abstract class SqlInternalOperators {
   public static final SqlInternalOperator GROUP_BY_DISTINCT =
       new SqlRollupOperator("GROUP BY DISTINCT", SqlKind.GROUP_BY_DISTINCT);
 
+  /** {@code GROUP BY ALL}, a placeholder expanded during validation into
+   * a standard {@code GROUP BY}. */
+  public static final SqlInternalOperator GROUP_BY_ALL =
+      new SqlRollupOperator("GROUP BY ALL", SqlKind.GROUP_BY_ALL);
+
   /** Fetch operator is ONLY used for its precedence during unparsing. */
   public static final SqlOperator FETCH =
       SqlBasicOperator.create("FETCH")

--- a/core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java
+++ b/core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java
@@ -5268,6 +5268,7 @@ public class SqlValidatorImpl implements SqlValidatorWithHints {
    * called even if no GROUP BY clause is present.
    */
   protected void validateGroupClause(SqlSelect select) {
+    rewriteGroupByAll(select);
     SqlNodeList groupList = select.getGroup();
     if (groupList == null) {
       return;
@@ -5337,6 +5338,32 @@ public class SqlValidatorImpl implements SqlValidatorWithHints {
     if (agg != null) {
       throw newValidationError(agg, RESOURCE.aggregateIllegalInClause(clause));
     }
+  }
+
+  /** If GROUP BY clause is the {@code GROUP BY ALL} placeholder, replaces it
+   * with every non-aggregated expression from the SELECT clause. */
+  private void rewriteGroupByAll(SqlSelect select) {
+    final SqlNodeList groupList = select.getGroup();
+    if (groupList == null
+        || groupList.size() != 1
+        || groupList.get(0).getKind() != SqlKind.GROUP_BY_ALL) {
+      return;
+    }
+    final List<SqlNode> keys = new ArrayList<>();
+    for (SqlNode selectItem : select.getSelectList()) {
+      if (SqlValidatorUtil.isMeasure(selectItem)) {
+        continue;
+      }
+      final SqlNode expr = SqlUtil.stripAs(selectItem);
+      if (expr instanceof SqlIdentifier && ((SqlIdentifier) expr).isStar()) {
+        throw newValidationError(expr,
+            RESOURCE.groupByAllRequiresExplicitSelectList());
+      }
+      if (aggOrOverFinder.findAgg(expr) == null) {
+        keys.add(expr);
+      }
+    }
+    select.setGroupBy(new SqlNodeList(keys, groupList.getParserPosition()));
   }
 
   private void validateGroupItem(SqlValidatorScope groupScope,

--- a/core/src/main/resources/org/apache/calcite/runtime/CalciteResource.properties
+++ b/core/src/main/resources/org/apache/calcite/runtime/CalciteResource.properties
@@ -131,6 +131,7 @@ GroupingInWrongClause={0} operator may only occur in SELECT, HAVING or ORDER BY 
 NotSelectDistinctExpr=Expression ''{0}'' is not in the select clause
 AggregateIllegalInClause=Aggregate expression is illegal in {0} clause
 WindowedAggregateIllegalInClause=Windowed aggregate expression is illegal in {0} clause
+GroupByAllRequiresExplicitSelectList=GROUP BY ALL requires an explicit SELECT list; ''*'' is not supported
 NestedAggIllegal=Aggregate expressions cannot be nested
 MeasureIllegal=Measure expressions can only occur within AGGREGATE function
 MeasureMustBeInAggregateQuery=Measure expressions can only occur within a GROUP BY query

--- a/core/src/test/java/org/apache/calcite/test/JdbcTest.java
+++ b/core/src/test/java/org/apache/calcite/test/JdbcTest.java
@@ -1297,6 +1297,17 @@ public class JdbcTest {
             + "c0=1998\n");
   }
 
+  /** Test case for [CALCITE-7594] GROUP BY ALL: grouping only by a constant
+   * over empty input returns 0 rows. */
+  @Test void testGroupByAllOverEmptyInput() {
+    CalciteAssert.hr()
+        .query("select 'x', count(*)\n"
+                + "from \"hr\".\"emps\"\n"
+                + "where false\n"
+                + "group by all")
+        .returnsCount(0);
+  }
+
   /** Test case for
    * <a href="https://issues.apache.org/jira/browse/CALCITE-2894">[CALCITE-2894]
    * NullPointerException thrown by RelMdPercentageOriginalRows when explaining

--- a/core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java
+++ b/core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java
@@ -7468,6 +7468,36 @@ public class SqlValidatorTest extends SqlValidatorTestCase {
         .withConformance(lenient).ok();
   }
 
+  @Test void testGroupByAll() {
+    // expands to every non-aggregate SELECT item - so listing a column besides
+    // an aggregate validates WITHOUT a "not being grouped" error
+    sql("select deptno, count(*) from emp group by all").ok();
+
+    // only aggregates -> global aggregation (one group), still valid
+    sql("select count(*) from emp group by all").ok();
+
+    // SELECT * cannot be expanded at group-validation time -> clear error
+    sql("select ^*^ from emp group by all")
+        .fails("(?s).*GROUP BY ALL requires an explicit SELECT list.*");
+
+    // contains-an-aggregate
+    sql("select deptno, substring(job, 1), count(*) + 1 as c, 'x' as x\n"
+        + "from emp group by all")
+        .ok();
+
+    //  GROUP BY ALL collects the non-aggregate exprs (sal, x + 1) as
+    // group keys and "x" still resolves, so the two features coexist.
+    sql("select sal as x, x + 1 as y, count(*) from emp group by all")
+        .withValidatorIdentifierExpansion(true)
+        .withConformance(SqlConformanceEnum.BABEL)
+        .ok();
+
+    // GROUP BY ALL is unaffected by isGroupByAlias
+    sql("select deptno as d, count(*) from emp group by all")
+        .withConformance(SqlConformanceEnum.LENIENT)
+        .ok();
+  }
+
   /** Test case for
    * <a href="https://issues.apache.org/jira/browse/CALCITE-5507">[CALCITE-5507]
    * HAVING alias failed when aggregate function in condition</a>. */

--- a/site/_docs/reference.md
+++ b/site/_docs/reference.md
@@ -210,7 +210,7 @@ select:
       [ BY expression [, expression ]* ]
       FROM tableExpression
       [ WHERE booleanExpression ]
-      [ GROUP BY [ ALL | DISTINCT ] { groupItem [, groupItem ]* } ]
+      [ GROUP BY { ALL | [ALL | DISTINCT ] groupItem [, groupiTEM ]* } ]
       [ HAVING booleanExpression ]
       [ WINDOW windowName AS windowSpec [, windowName AS windowSpec ]* ]
       [ QUALIFY booleanExpression ]
@@ -422,7 +422,12 @@ may refer to tables in the FROM clause of an enclosing query.
 GROUP BY DISTINCT removes duplicate grouping sets (for example,
 "GROUP BY DISTINCT GROUPING SETS ((a), (a, b), (a))" is equivalent to
 "GROUP BY GROUPING SETS ((a), (a, b))");
-GROUP BY ALL is equivalent to GROUP BY.
+GROUP BY ALL followed by grouping items is equivalent to GROUP BY
+(ALL is the default set quantifier).
+GROUP BY ALL on its own groups by every expression in the SELECT clause
+that is not an aggregate function; for example,
+"SELECT deptno, SUM(sal) FROM emp GROUP BY ALL" is equivalent to
+"SELECT deptno, SUM(sal) FROM emp GROUP BY deptno".
 
 *selectWithoutFrom* is equivalent to VALUES,
 but is not standard SQL and is only allowed in certain

--- a/testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java
+++ b/testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java
@@ -2637,6 +2637,22 @@ public class SqlParserTest {
     sql(sql2).ok(expected2);
   }
 
+  @Test void testGroupByAll() {
+    final String sql = "select x, sum(y) from t\n"
+        + "group by all";
+    final String expected = "SELECT `X`, SUM(`Y`)\n"
+        + "FROM `T`\n"
+        + "GROUP BY ALL";
+    sql(sql).ok(expected);
+
+    final String sql1 = "select deptno from emp\n"
+        + "group by all deptno, gender";
+    final String expected1 = "SELECT `DEPTNO`\n"
+        + "FROM `EMP`\n"
+        + "GROUP BY `DEPTNO`, `GENDER`";
+    sql(sql1).ok(expected1);
+  }
+
   @Test void testGroupByCube2() {
     final String sql = "select deptno from emp\n"
         + "group by cube ((a, b), (c, d)) order by a";

